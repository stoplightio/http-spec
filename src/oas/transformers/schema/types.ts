import type { JSONSchema7 } from 'json-schema';
import type { SchemaObject } from 'openapi3-ts/src/model/OpenApi';
import type { Schema } from 'swagger-schema-official';

export type OASSchemaObject = SchemaObject | Schema;

export type Converter = (schema: OASSchemaObject | JSONSchema7) => void;
