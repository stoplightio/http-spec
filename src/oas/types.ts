import { DeepPartial, IHttpOperation, IHttpService } from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

export const enum OasVersion {
  OAS2 = 2,
  OAS3 = 3,
}

/**
 * Service
 */

export interface ITransformServiceOpts<T> {
  document: DeepPartial<T>;
}

export type HttpServiceTransformer<T> = (opts: T) => IHttpService;

export type Oas2TransformServiceOpts = ITransformServiceOpts<Spec>;
export type Oas3TransformServiceOpts = ITransformServiceOpts<OpenAPIObject>;
export type Oas2HttpServiceTransformer = HttpServiceTransformer<Oas2TransformServiceOpts>;
export type Oas3HttpServiceTransformer = HttpServiceTransformer<Oas3TransformServiceOpts>;

/**
 * Operation
 */

export interface ITransformOperationOpts<T> {
  document: DeepPartial<T>;
  path: string;
  method: string;
}

export type HttpOperationTransformer<T> = (opts: T) => IHttpOperation;

export type Oas2TransformOperationOpts = ITransformOperationOpts<Spec>;
export type Oas3TransformOperationOpts = ITransformOperationOpts<OpenAPIObject>;
export type Oas2HttpOperationTransformer = HttpOperationTransformer<Oas2TransformOperationOpts>;
export type Oas3HttpOperationTransformer = HttpOperationTransformer<Oas3TransformOperationOpts>;
