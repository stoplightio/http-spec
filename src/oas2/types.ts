import type { DeepPartial } from '@stoplight/types';
import type { Spec } from 'swagger-schema-official';

import { TranslateFunction } from '../types';

export type Oas2TranslateFunction<P extends unknown[], R extends unknown = unknown> = TranslateFunction<
  DeepPartial<Spec>,
  P,
  R
>;
