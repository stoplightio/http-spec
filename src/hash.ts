// @ts-expect-error: no types
import * as fnv from 'fnv-plus';

import type { IdGenerator } from './types';

// export const hash: IdGenerator = input => fnv.fast1a52hex(input);
export const hash: IdGenerator = input => input;
