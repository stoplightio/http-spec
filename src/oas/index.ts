export * from './operation';
export * from './types';
