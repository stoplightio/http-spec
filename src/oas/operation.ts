import { isPlainObject } from '@stoplight/json';
import type { DeepPartial, IHttpEndpointOperation } from '@stoplight/types';
import type { OpenAPIObject, OperationObject, PathsObject } from 'openapi3-ts';
import type { Spec } from 'swagger-schema-official';
import pickBy = require('lodash.pickby');

import { isBoolean, isString } from '../guards';
import type { EndpointOperationConfig, Fragment, HttpEndpointOperationTransformer } from '../types';
import { TransformerContext, TranslateFunction } from '../types';
import { extractId } from '../utils';
import { getExtensions } from './accessors';
import { toExternalDocs } from './externalDocs';
import { translateToTags } from './tags';
import { translateToSecurityDeclarationType } from './transformers';

const DEFAULT_METHODS = ['get', 'post', 'put', 'delete', 'options', 'head', 'patch', 'trace'];

export const OPERATION_CONFIG: EndpointOperationConfig = {
  type: 'operation',
  documentProp: 'paths',
  nameProp: 'path',
};

export const WEBHOOK_CONFIG: EndpointOperationConfig = {
  type: 'webhook',
  documentProp: 'webhooks',
  nameProp: 'name',
};

export function transformOasEndpointOperations<
  T extends Fragment & DeepPartial<Spec | OpenAPIObject>,
  TEndpoint extends IHttpEndpointOperation,
>(
  document: T,
  transformer: HttpEndpointOperationTransformer<any, TEndpoint>,
  config: EndpointOperationConfig,
  methods: string[] | null = DEFAULT_METHODS,

  ctx?: TransformerContext<T>,
): TEndpoint[] {
  const entries = isPlainObject(document[config.documentProp]) ? Object.entries(document[config.documentProp]) : [];

  return entries.flatMap(([name, value]) => {
    if (!isPlainObject(value)) return [];

    let operations = Object.keys(value);
    if (methods !== null) {
      operations = operations.filter(pathKey => methods.includes(pathKey));
    }

    return operations.map(method =>
      transformer({
        document,
        name,
        method,
        config,
        ctx,
      }),
    );
  });
}

export const transformOasEndpointOperation: TranslateFunction<
  DeepPartial<OpenAPIObject> | DeepPartial<Spec>,
  [config: EndpointOperationConfig, name: string, method: string],
  Omit<IHttpEndpointOperation, 'responses' | 'request' | 'servers' | 'security' | 'callbacks'>
> = function ({ type, documentProp, nameProp }: EndpointOperationConfig, name: string, method: string) {
  const pathObj = this.maybeResolveLocalRef(this.document?.[documentProp]?.[name]) as PathsObject;
  if (typeof pathObj !== 'object' || pathObj === null) {
    throw new Error(`Could not find ${[documentProp, name].join('/')} in the provided spec.`);
  }

  const obj = this.maybeResolveLocalRef(pathObj[method]) as OperationObject;
  if (!obj) {
    throw new Error(`Could not find ${[documentProp, name, method].join('/')} in the provided spec.`);
  }

  const serviceId = (this.ids.service = String(this.document['x-stoplight']?.id));
  if (type === 'operation') {
    this.ids.path = this.generateId.httpPath({ parentId: serviceId, path: name });
  } else {
    this.ids.webhookName = this.generateId.httpWebhookName({ parentId: serviceId, name });
  }
  let id: string;
  if (this.context === 'callback') {
    id = this.ids.operation =
      extractId(obj) ??
      this.generateId.httpCallbackOperation({
        parentId: serviceId,
        method,
        path: name,
      });
  } else if (type === 'operation') {
    id = this.ids.operation =
      extractId(obj) ?? this.generateId.httpOperation({ parentId: serviceId, method, path: name });
  } else {
    id = this.ids.webhook = extractId(obj) ?? this.generateId.httpWebhook({ parentId: serviceId, method, name });
  }

  this.parentId = id;

  this.context = type;

  return {
    id,

    method,
    [nameProp]: name,

    tags: translateToTags.call(this, obj.tags),
    extensions: getExtensions(obj),

    ...pickBy(
      {
        deprecated: obj.deprecated,
        internal: obj['x-internal'],
      },
      isBoolean,
    ),

    ...pickBy(
      {
        iid: obj.operationId,
        description: obj.description,
        summary: obj.summary,
      },
      isString,
    ),

    securityDeclarationType: translateToSecurityDeclarationType(obj),
    ...toExternalDocs(obj.externalDocs),
  };
};
