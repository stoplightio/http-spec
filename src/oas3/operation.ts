import type { DeepPartial, IHttpOperation } from '@stoplight/types';
import pickBy from 'lodash.pickby';
import type { OpenAPIObject } from 'openapi3-ts';

import { createContext, DEFAULT_ID_GENERATOR } from '../context';
import { isNonNullable } from '../guards';
import { transformOasOperation, transformOasOperations } from '../oas';
import { resolveRef } from '../oas/resolver';
import type { Oas3HttpOperationTransformer } from '../oas/types';
import { Fragment } from '../types';
import { translateToCallbacks } from './transformers/callbacks';
import { translateToRequest } from './transformers/request';
import { translateToResponses } from './transformers/responses';
import { translateToSecurities } from './transformers/securities';
import { translateToServers } from './transformers/servers';

export function transformOas3Operations(document: DeepPartial<OpenAPIObject>): IHttpOperation[] {
  return transformOasOperations(document, transformOas3Operation);
}

export const transformOas3Operation: Oas3HttpOperationTransformer = ({ document: _document, path, method }) => {
  const ctx = createContext(_document, resolveRef, DEFAULT_ID_GENERATOR);
  const httpOperation = transformOasOperation.call(ctx, path, method);
  const pathObj = ctx.maybeResolveLocalRef(ctx.document.paths![path]) as Fragment;
  const operation = ctx.maybeResolveLocalRef(pathObj[method]) as Fragment;

  return {
    ...httpOperation,

    responses: translateToResponses.call(ctx, operation.responses),
    request: translateToRequest.call(ctx, pathObj, operation),
    security: translateToSecurities.call(ctx, operation.security),
    servers: translateToServers.call(ctx, pathObj, operation),

    ...pickBy(
      {
        callbacks: translateToCallbacks.call(ctx, operation.callbacks),
      },
      isNonNullable,
    ),
  };
};
