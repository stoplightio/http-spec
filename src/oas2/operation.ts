import type { DeepPartial, IHttpOperation } from '@stoplight/types';
import type { Spec } from 'swagger-schema-official';

import { createContext } from '../oas/context';
import { OPERATION_CONFIG, transformOasEndpointOperation, transformOasEndpointOperations } from '../oas/operation';
import { Oas2HttpOperationTransformer } from '../oas/types';
import type { Fragment } from '../types';
import { TransformerContext } from '../types';
import { translateToRequest } from './transformers/request';
import { translateToResponses } from './transformers/responses';
import { translateToSecurities } from './transformers/securities';
import { translateToServers } from './transformers/servers';

export function transformOas2Operations<T extends DeepPartial<Spec> = DeepPartial<Spec>>(
  document: T,
  ctx?: TransformerContext<T>,
): IHttpOperation[] {
  return transformOasEndpointOperations(document, transformOas2Operation, OPERATION_CONFIG, void 0, ctx);
}

export const transformOas2Operation: Oas2HttpOperationTransformer = ({
  document: _document,
  name,
  method,
  config,
  ctx = createContext(_document),
}) => {
  const httpEndpointOperation = transformOasEndpointOperation.call(ctx, config, name, method);
  const parentObj = ctx.maybeResolveLocalRef(ctx.document[config.documentProp]![name]) as Fragment;
  const obj = ctx.maybeResolveLocalRef(parentObj[method]) as Fragment;

  return {
    ...httpEndpointOperation,

    responses: translateToResponses.call(ctx, obj),
    servers: translateToServers.call(ctx, obj),
    request: translateToRequest.call(ctx, parentObj, obj),
    security: translateToSecurities.call(ctx, obj.security, 'requirement'),
  } as any;
};
