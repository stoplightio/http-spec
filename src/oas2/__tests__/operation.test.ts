import { DeepPartial } from '@stoplight/types';
import { Spec } from 'swagger-schema-official';

import { setSkipHashing } from '../../hash';
import { OPERATION_CONFIG } from '../../oas/operation';
import { transformOas2Operation, transformOas2Operations } from '../operation';

setSkipHashing(true);

describe('transformOas2Operation', () => {
  it('should translate operation', () => {
    const document: Partial<Spec> = {
      swagger: '2.0',
      info: {
        title: 'title',
        version: '1.0',
      },
      paths: {
        '/users/{userId}': {
          get: {
            operationId: 'oid',
            description: 'odesc',
            deprecated: true,
            summary: 'osum',
            responses: {
              response: {
                description: 'desc',
                examples: {
                  example: {},
                },
                headers: {
                  header: {
                    type: 'integer',
                  },
                },
                schema: {},
              },
            },
            security: [
              {
                petstore_auth: ['write:pets'],
                api_key: [],
              },
              {
                api_key: [],
              },
              {},
            ],
          },
        },
      },
      securityDefinitions: {
        petstore_auth: {
          type: 'oauth2',
          authorizationUrl: 'https://petstore.swagger.io/oauth/dialog',
          flow: 'implicit',
          scopes: {
            'write:pets': 'modify pets in your account',
            'read:pets': 'read your pets',
          },
        },
        api_key: {
          type: 'apiKey',
          name: 'api_key_name',
          in: 'header',
        },
      },
    };

    expect(
      transformOas2Operation({
        name: '/users/{userId}',
        method: 'get',
        document,
        config: OPERATION_CONFIG,
      }),
    ).toMatchSnapshot({
      id: expect.any(String),
      responses: [
        {
          id: expect.any(String),
          contents: [
            {
              id: expect.any(String),
              examples: [
                {
                  id: expect.any(String),
                },
              ],
            },
          ],
          headers: [
            {
              id: expect.any(String),
              schema: {
                'x-stoplight': {
                  id: expect.any(String),
                },
              },
            },
          ],
        },
      ],
      security: [
        [
          {
            id: expect.any(String),
            extensions: {},
          },
          {
            id: expect.any(String),
            extensions: {},
          },
        ],
        [
          {
            id: expect.any(String),
            extensions: {},
          },
        ],
        [],
      ],
    });
  });

  it('should properly translate operation with no response body', () => {
    const document: Partial<Spec> = {
      swagger: '2.0',
      info: {
        title: 'title',
        version: '1.0',
      },
      produces: ['application/json'],
      paths: {
        '/users/{userId}': {
          delete: {
            operationId: 'oid',
            description: 'odesc',
            summary: 'osum',
            responses: {
              '204': {
                description: 'Some description',
              },
            },
          },
        },
      },
    };

    const result = transformOas2Operation({
      name: '/users/{userId}',
      method: 'delete',
      document,
      config: OPERATION_CONFIG,
    });

    expect(result.responses[0].contents).toHaveLength(0);
  });

  it('should return deprecated property in http operation root', () => {
    const document: DeepPartial<Spec> = {
      swagger: '2.0',
      paths: {
        '/users/{userId}': {
          get: {
            deprecated: true,
          },
        },
      },
    };

    expect(
      transformOas2Operation({
        name: '/users/{userId}',
        method: 'get',
        document,
        config: OPERATION_CONFIG,
      }),
    ).toHaveProperty('deprecated', true);
  });

  it('should return x-internal property in http operation root', () => {
    const document: DeepPartial<Spec & { paths: any }> = {
      swagger: '2.0',
      paths: {
        '/users/{userId}': {
          get: {
            'x-internal': true,
          },
          post: {
            'x-internal': false,
          },
          put: {},
        },
      },
    };

    expect(transformOas2Operations(document as any)).toStrictEqual([
      expect.objectContaining({
        path: '/users/{userId}',
        method: 'get',
        internal: true,
        extensions: {},
      }),
      expect.objectContaining({
        path: '/users/{userId}',
        method: 'post',
        internal: false,
        extensions: {},
      }),
      {
        id: expect.any(String),
        path: '/users/{userId}',
        method: 'put',
        request: {
          cookie: [],
          headers: [],
          path: [],
          query: [],
        },
        responses: [],
        security: [],
        securityDeclarationType: 'inheritedFromService',
        servers: [],
        tags: [],
        extensions: {},
      },
    ]);
  });

  it('given malformed parameters should translate operation with those parameters', () => {
    const document: DeepPartial<Spec> = {
      swagger: '2.0',
      paths: {
        '/users/{userId}': {
          get: {
            parameters: [
              {
                in: 'header',
                name: 'name',
              },
              null as any,
            ],
          },
        },
      },
    };

    expect(
      transformOas2Operation({
        name: '/users/{userId}',
        method: 'get',
        document,
        config: OPERATION_CONFIG,
      }),
    ).toStrictEqual({
      id: expect.any(String),
      method: 'get',
      path: '/users/{userId}',
      request: {
        cookie: [],
        headers: [
          {
            id: expect.any(String),
            name: 'name',
            style: 'simple',
          },
        ],
        path: [],
        query: [],
      },
      responses: [],
      security: [],
      securityDeclarationType: 'inheritedFromService',
      servers: [],
      tags: [],
      extensions: {},
    });
  });
});
