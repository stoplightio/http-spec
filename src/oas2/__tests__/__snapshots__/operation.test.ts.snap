// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformOas2Operation should translate operation 1`] = `
Object {
  "deprecated": true,
  "description": "odesc",
  "id": "?http-operation-id?",
  "iid": "oid",
  "method": "get",
  "path": "/users/{userId}",
  "request": Object {},
  "responses": Array [
    Object {
      "code": "response",
      "contents": Array [
        Object {
          "examples": Array [
            Object {
              "key": "example",
              "value": Object {},
            },
          ],
          "mediaType": "",
          "schema": Object {},
        },
      ],
      "description": "desc",
      "headers": Array [
        Object {
          "description": undefined,
          "name": "header",
          "schema": Object {
            "type": "integer",
          },
          "style": "simple",
        },
      ],
    },
  ],
  "security": Array [
    Array [
      Object {
        "description": undefined,
        "flows": Object {
          "implicit": Object {
            "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
            "scopes": Object {
              "write:pets": "modify pets in your account",
            },
          },
        },
        "key": "petstore_auth",
        "type": "oauth2",
      },
    ],
    Array [
      Object {
        "description": undefined,
        "in": "header",
        "key": "api_key",
        "name": "api_key_name",
        "type": "apiKey",
      },
    ],
  ],
  "servers": Array [],
  "summary": "osum",
  "tags": Array [],
}
`;
